ext/ldap/ldap.c

```diff
@@ -1593,20 +1737,25 @@ static void php_ldap_do_search(INTERNAL_FUNCTION_PARAMETERS, int scope)
 cleanup_parallel:
 		efree(lds);
 		efree(rcs);
-	} else {
-		ld = (ldap_linkdata *) zend_fetch_resource_ex(link, "ldap link", le_link);
-		if (ld == NULL) {
+	} else if (Z_TYPE_P(link) == IS_OBJECT && instanceof_function(Z_OBJCE_P(link), ldap_link_ce)) {
+		ld = Z_LDAP_LINK_P(link);
+		if (!ld->link) {
+			zend_throw_error(NULL, "LDAP connection has already been closed");
 			ret = 0;
 			goto cleanup;
 		}
 
 		if (!base_dn_str) {
-			zend_argument_type_error(2, "must be of type string when argument #1 ($ldap) is a resource");
+			zend_argument_type_error(2, "must be of type string when argument #1 ($ldap) is an LDAPLink");
+			ret = 0;
```

Comment:
I think it used to be a bug that these 2 lines were missing. If someone can verify it, I'll apply this change to earlier branches (and the one below).

Reply:
Right, previously this would crash if base_dn_str / filter_str were actually null. I'd suggest to add a test for this as well.